### PR TITLE
Add streaming logs from device agents

### PR DESCRIPTION
**Pull request by hardillb**
## Description
This requires MQTT device comms

Dependent on https://github.com/flowforge/flowforge-device-agent/pull/66

## Related Issue(s)

https://github.com/flowforge/flowforge/issues/1820

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `flowforge/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `flowforge/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

File: frontend/src/pages/device/components/DeviceLog.vue [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    },
+    unmounted () {
+        // need to unsubscribe here
+        this.disconnect()
+        clearInterval(this.keepAliveInterval)
+    },
+    methods: {
+        connect: async function () {
+            await DeviceApi.startLogs(this.device.id)
+            console.log('connect')
```
Comment (Steve-Mcl) [inline frontend/src/pages/device/components/DeviceLog.vue:53]:
```suggestion
```

File: frontend/src/pages/device/components/DeviceLog.vue [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            // this.keepAliveInterval = setInterval(() => {
+            //     this.connect(this.device.id)
+            // }, 10000)
+            if (this.connection === null) {
+                try {
+                    const protocol = location.protocol === 'http:' ? 'ws:' : 'wss:'
+                    console.log(`connecting to ${protocol}//${location.host}/api/v1/devices/${this.device.id}/logs`)
```
Comment (Steve-Mcl) [inline frontend/src/pages/device/components/DeviceLog.vue:60]:
```suggestion
```

File: frontend/src/pages/device/components/DeviceLog.vue [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                    this.connection.onopen = () => {
+                        this.loading = false
+                    }
+                    this.connection.onmessage = message => {
+                        const m = JSON.parse(message.data)
+                        if (!Array.isArray(m)) {
+                            if (!isNaN(m.ts)) {
+                                m.ts = `${m.ts}`
+                            }
+                            console.log(m)
```
Comment (Steve-Mcl) [inline frontend/src/pages/device/components/DeviceLog.vue:71]:
```suggestion
```

File: frontend/src/pages/device/components/DeviceLog.vue [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                            const d = new Date(parseInt(m.ts.substring(0, m.ts.length - 4)))
+                            m.date = `${d.toLocaleDateString()} ${d.toLocaleTimeString()}`
+                            this.logEntries.push(m)
+                        } else {
+                            console.log('array')
```
Comment (Steve-Mcl) [inline frontend/src/pages/device/components/DeviceLog.vue:76]:
```suggestion
```

File: frontend/src/pages/device/components/DeviceLog.vue [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                        }
+                    }
+                } catch (e) {
+                    console.log(e)
+                }
+            }
+        },
+        disconnect: async function () {
+            DeviceApi.stopLogs(this.device.id)
+            console.log('disconnect')
```
Comment (Steve-Mcl) [inline frontend/src/pages/device/components/DeviceLog.vue:94]:
```suggestion
```

File: frontend/src/pages/device/index.vue [hunk 6]
```diff
@@ -104,6 +106,17 @@ export default {
             const device = await deviceApi.getDevice(this.$route.params.id)
             this.device = device
             this.$store.dispatch('account/setTeam', this.device.team.slug)
+        },
+        checkFeatures: async function () {
+            console.log(this.features)
```
Comment (Steve-Mcl) [inline frontend/src/pages/device/index.vue:111]:
```suggestion
```

File: frontend/src/pages/device/Logs.vue [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<template>
+    <SectionTopMenu hero="Device Logs" help-header="FlowForge - Device Logs" info="Live logs from your FlowForge instances of Node-RED">
+        <!-- <template>
+        </template>
+        <template #tools>
+        </template> -->
```
Comment (Steve-Mcl) [inline frontend/src/pages/device/Logs.vue:6]:
```suggestion
```

Should remove this?

File: frontend/src/api/devices.js [hunk 8]
```diff
@@ -52,6 +52,10 @@ const updateSettings = async (deviceId, settings) => {
     })
 }
 
+const startLogs = async (deviceId) => {}
+
+const stopLogs = async (deviceId) => {}
+
```
Comment (Steve-Mcl) [inline frontend/src/api/devices.js:58]: Are these for future use Ben?

File: frontend/src/pages/device/components/DeviceLog.vue [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        // need to unsubscribe here
+        this.disconnect()
+        clearInterval(this.keepAliveInterval)
+    },
+    methods: {
+        connect: async function () {
+            await DeviceApi.startLogs(this.device.id)
+            // this.keepAliveInterval = setInterval(() => {
+            //     this.connect(this.device.id)
+            // }, 10000)
```
Comment (Steve-Mcl) [inline frontend/src/pages/device/components/DeviceLog.vue:55]:
```suggestion
```

Remove unused code

File: frontend/src/api/devices.js (hunk 8 above)
Comment (hardillb) [reply frontend/src/api/devices.js:58]:
```suggestion
```

They should be obsolete now, the counter in the websocket code should handle all that now.

Comment (Steve-Mcl) [reply frontend/src/api/devices.js:58]: ok, which means the `exports` at lines 68,69 and the async call in https://github.com/flowforge/flowforge/blob/365bf107be8c8b46cc3c101924469fc44019fb13/frontend/src/pages/device/components/DeviceLog.vue#L52 should also be removed right?

File: frontend/src/pages/device/components/DeviceLog.vue [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        this.connect()
+    },
+    unmounted () {
+        // need to unsubscribe here
+        this.disconnect()
+        clearInterval(this.keepAliveInterval)
+    },
+    methods: {
+        connect: async function () {
+            await DeviceApi.startLogs(this.device.id)
```
Comment (hardillb) [inline frontend/src/pages/device/components/DeviceLog.vue:52]:
```suggestion
```

File: frontend/src/pages/device/components/DeviceLog.vue [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+                            })
+                        }
+                    }
+                } catch (e) {
+                    console.error(e)
+                }
+            }
+        },
+        disconnect: async function () {
+            DeviceApi.stopLogs(this.device.id)
```
Comment (hardillb) [inline frontend/src/pages/device/components/DeviceLog.vue:86]:
```suggestion
```

File: frontend/src/api/devices.js [hunk 12]
```diff
@@ -60,5 +64,7 @@ export default {
     updateDevice,
     generateCredentials,
     getSettings,
-    updateSettings
+    updateSettings,
+    startLogs,
+    stopLogs
```
Comment (hardillb) [inline frontend/src/api/devices.js:69]:
```suggestion
    updateSettings
```

File: frontend/src/pages/device/Logs.vue (hunk 7 above)
Comment (Steve-Mcl) [reply frontend/src/pages/device/Logs.vue:6]: last one @hardillb 

Is it necessary to keep commented code?

Comment (hardillb) [reply frontend/src/pages/device/Logs.vue:6]: I don't think so, I removed them when talking to Pez about the lint changes. I'll commit this suggestion.

Please pull and re-test this before merging just to be sure

**Review by Steve-Mcl — APPROVED** [review]
All good for me.  Will merge once tests complete.

**Pull request merged**